R/R/compile.R

```diff
@@ -0,0 +1,110 @@
+    }
+    makefile <- file.path(folder, "Makefile")
+    if (!file.exists(makefile)) {
+        stop(paste0("BridgeStan folder '", folder, "' does not contain file 'Makefile',",
+            " please ensure it is built properly!\n", "If you need to set a different location, call 'set_bridgestan_path()'"))
+    }
+}
+
+#' Set the path to BridgeStan.
+#'
+#' This should point to the top-level folder of the repository.
+#' @export
+set_bridgestan_path <- function(path) {
+    verify_bridgestan_path(path)
+    Sys.setenv(BRIDGESTAN = normalizePath(path))
+}
+
+get_bridgestan_path <- function() {
+    # try to get from environment
+    path <- Sys.getenv("BRIDGESTAN", unset = "")
+    if (path == "") {
+        path <- CURRENT_BRIDGESTAN
+        tryCatch({
+            verify_bridgestan_path(path)
+        }, error = function(e) {
+            print(paste0("Bridgestan not found at location specified by $BRIDGESTAN ",
+                "environment variable, downloading version ", packageVersion("bridgestan"),
+                " to ", path))
+            get_bridgestan_src()
+        })
+    }
+
+    return(path)
+}
+
+
+#' Run BridgeStan's Makefile on a `.stan` file, creating the `.so`
+#' used by the StanModel class.
+#' This function assumes that the path to BridgeStan is valid.
+#' This can be set with `set_bridgestan_path`.
+#'
+#' @param stan_file A path to a Stan model file.
+#' @param stanc_arg A list of arguments to pass to stanc3.
+#' For example, `c('--O1')` will enable compiler optimization level 1.
+#' @param make_args A list of additional arguments to pass to Make.
+#' For example, `c('STAN_THREADS=True')` will enable
+#' threading for the compiled model. If the same flags are defined
+#' in `make/local`, the versions passed here will take precedent.
+#' @return Path to the compiled model.
+#' @export
+compile_model <- function(stan_file, stanc_args = NULL, make_args = NULL) {
+    verify_bridgestan_path(get_bridgestan_path())
+    suppressWarnings({
+        file_path <- normalizePath(stan_file)
+    })
+    if (tools::file_ext(file_path) != "stan") {
+        stop(paste0("File '", file_path, "' does not end with '.stan'"))
+    }
+    if (!file.exists(file_path)) {
+        stop(paste0("File '", file_path, "' does not exist!"))
+    }
+
+    output <- paste0(tools::file_path_sans_ext(file_path), "_model.so")
+    stancflags <- paste("--include-paths=.", paste(stanc_args, collapse = " "))
+
+    flags <- c(paste("-C", get_bridgestan_path()), make_args, paste0("STANCFLAGS=\"",
+        stancflags, "\""), output)
+
+    suppressWarnings({
+        res <- system2(MAKE, args = flags, stdout = TRUE, stderr = TRUE)
+    })
+    res_attrs <- attributes(res)
+    if ("status" %in% names(res_attrs) && res_attrs$status != 0) {
+        stop(paste0("Compilation failed with error code ", res_attrs$status, "\noutput:\n",
+            paste(res, collapse = "\n")))
+    }
+
+    return(output)
+}
+
+windows_path_setup <- function() {
```

Comment:
I missed it.  Where is this function used?

Reply:
It's called in the constructor before we load the shared object (same as Python, looks like Julia calls it at startup which is not ideal and I will take a look at)